File: packages/python-sdk/src/agentic_crm_sdk/leads.py
"""
Leads API client module.
"""
from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from agentic_crm_sdk.async_client import AsyncAgenticCRM

from agentic_crm_sdk.client import AgenticCRM


class LeadsClient:
    def __init__(self, client: AgenticCRM):
        self._client = client

    def list(self, **params: Any) -> Any:
        return self._client.get("/api/leads", params=params)

    def get(self, lead_id: int) -> Any:
        return self._client.get(f"/api/leads/{lead_id}")

    def create(self, data: dict) -> Any:
        return self._client.post("/api/leads", json=data)

    def update(self, lead_id: int, data: dict) -> Any:
        return self._client.patch(f"/api/leads/{lead_id}", json=data)

    def delete(self, lead_id: int) -> Any:
        return self._client.delete(f"/api/leads/{lead_id}")

    def activities(self, lead_id: int) -> Any:
        return self._client.get(f"/api/leads/{lead_id}/activities")

    def add_note(self, lead_id: int, content: str) -> Any:
        return self._client.post(f"/api/leads/{lead_id}/notes", json={"content": content})


class AsyncLeadsClient:
    def __init__(self, client: AsyncAgenticCRM):
        self._crm = client

    async def list(self, **params: Any) -> Any:
        return await self._crm.get("/api/leads", params=params)

    async def get(self, lead_id: int) -> Any:
        return await self._crm.get(f"/api/leads/{lead_id}")

    async def create(self, data: dict) -> Any:
        return await self._crm.post("/api/leads", json=data)

    async def update(self, lead_id: int, data: dict) -> Any:
        return await self._crm.patch(f"/api/leads/{lead_id}", json=data)

    async def delete(self, lead_id: int) -> Any:
        return await self._crm.delete(f"/api/leads/{lead_id}")

    async def activities(self, lead_id: int) -> Any:
        return await self._crm.get(f"/api/leads/{lead_id}/activities")

    async def add_note(self, lead_id: int, content: str) -> Any:
        return await self._crm.post(f"/api/leads/{lead_id}/notes", json={"content": content})

File: frontend/lib/sdk/leads.ts
import type { Lead, LeadSource, LeadStage } from '../api'
import { AgenticCRM } from './client'

export interface CreateLeadInput {
  contact_id: number
  source?: LeadSource
  stage?: LeadStage
  score?: number
  tags?: string[]
  notes?: string
  assigned_agent_id?: number
}

export interface UpdateLeadInput {
  source?: LeadSource
  stage?: LeadStage
  score?: number
  tags?: string[]
  notes?: string
  assigned_agent_id?: number
  last_contacted_at?: string
}

export class LeadsClient {
  constructor(private client: AgenticCRM) {}

  async list(params?: { stage?: LeadStage; source?: LeadSource; search?: string }): Promise<Lead[]> {
    return this.client.get<Lead[]>('/api/leads', params as Record<string, string>)
  }

  async get(leadId: number): Promise<Lead> {
    return this.client.get<Lead>(`/api/leads/${leadId}`)
  }

  async create(data: CreateLeadInput): Promise<Lead> {
    return this.client.post<Lead>('/api/leads', data)
  }

  async update(leadId: number, data: UpdateLeadInput): Promise<Lead> {
    return this.client.patch<Lead>(`/api/leads/${leadId}`, data)
  }

  async delete(leadId: number): Promise<void> {
    return this.client.delete<void>(`/api/leads/${leadId}`)
  }

  async activities(leadId: number): Promise<unknown[]> {
    return this.client.get<unknown[]>(`/api/leads/${leadId}/activities`)
  }

  async addNote(leadId: number, content: string): Promise<void> {
    return this.client.post<void>(`/api/leads/${leadId}/notes`, { content })
  }
}